DOCSP-42349: Transactions (#89)
mongodb/docs-scala: source/write/transactions.txt

transaction:

.. list-table::
   :widths: 25 75
   :stub-columns: 1
   :header-rows: 1

   * - Method
     - Description

   * - ``startTransaction()``
     - | Starts a new transaction on this session. You cannot start a
         transaction if there's already an active transaction running in
         the session.
       |
       | You can set transaction options by passing a ``TransactionOptions``
         instance as a parameter.

   * - ``commitTransaction()``
     - | Commits the active transaction for this session. This method returns an
         error if there is no active transaction for the session, the
         transaction was previously ended, or if there is a write conflict.

   * - ``abortTransaction()``
     - | Ends the active transaction for this session. This method returns an
         error if there is no active transaction for the session or if the
         transaction was committed or ended.

.. _scala-txn-example:

Transaction Example
-------------------

This example defines a ``runTransaction()`` method that
modifies data in the collections of the ``sample_mflix`` database.
The code performs the following actions:

> - Creates ``MongoCollection`` instances to access the ``movies``
>   and ``users`` collections.
> - Specifies the read preference, read concern, and write concern for
>   the transaction.
> - Starts the transaction.
> - Inserts a document into the ``movies`` collection and prints the results.
> - Updates a document in the ``users`` collection and prints the results.

Reviewer: Since these aren't full sentences
```suggestion
- Creates ``MongoCollection`` instances to access the ``movies``
  and ``users`` collections
- Specifies the read preference, read concern, and write concern for
  the transaction
- Starts the transaction
- Inserts a document into the ``movies`` collection and prints the results
- Updates a document in the ``users`` collection and prints the results
```

.. literalinclude:: /includes/write/transaction.scala
   :copyable:
   :language: scala
   :dedent:
   :start-after: begin-transaction-method
   :end-before: end-transaction-method

Then, run the following code to perform the transaction. This code
completes the following actions:

> - Creates a session from the client by using the ``startSession()`` method.
> - Calls the ``runTransaction()`` method defined in the preceding example, passing
>   the database and the session as parameters.
> - Commits the transaction by calling the ``commitTransaction()`` method and waits
>   for the operations to complete.

Reviewer: ```suggestion
- Creates a session from the client by using the ``startSession()`` method
- Calls the ``runTransaction()`` method defined in the preceding example, passing
  the database and the session as parameters
- Commits the transaction by calling the ``commitTransaction()`` method and waits
  for the operations to complete
```

.. io-code-block::
   :copyable:

   .. input:: /includes/write/transaction.scala
      :language: scala
      :dedent:
      :start-after: start-perform-transaction
      :end-before: end-perform-transaction

   .. output:: 
      :language: console
      :visible: false

      AcknowledgedUpdateResult{matchedCount=1, modifiedCount=1, upsertedId=null}
      AcknowledgedInsertOneResult{insertedId=BsonObjectId{value=...}}

Additional Information
----------------------

To learn more about the concepts mentioned in this guide, see the
following pages in the {+mdb-server+} manual:

- :manual:`Transactions </core/transactions/>`
- :manual:`Server Sessions </reference/server-sessions/>`
- :manual:`Read Isolation, Consistency, and Recency
  </core/read-isolation-consistency-recency/>`

> To learn more about ACID compliance, see the :website:`What are ACID
> Properties in Database Management Systems? </basics/acid-transactions>`

Reviewer: Looks like the article name changed
```suggestion
To learn more about ACID compliance, see the :website:`A Guide to ACID Properties in Database Management Systems </basics/acid-transactions>`
```
